add: Deployment using serverless-webpack in lambda

danieluhm2004 opened the pull request:
## Describe this PR

When deploying using serverless-webpack , it is impossible to deploy except for a specific engine using package.pattern as in the documentation. So, the Lambda engine capacity reached almost 80MB, which made it impossible to deploy.

## Changes

- The serverless-webpack-prisma library was developed separately to make it easier to use.
- Documentation has been added so that all users who use serverless-webpack can use Prisma smoothly.

## What issue does this fix?

- https://github.com/prisma/prisma/issues/6644
- https://github.com/prisma/prisma/issues/2303

## Any other relevant information

None!

janpio commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,35 @@ package:
 
 This ensures the packaged archive is as small as possible.
 
+### Deployment using ``serverless-webpack``
```
```suggestion
### Deployment using `serverless-webpack``
```

janpio commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,35 @@ package:
 
+### Deployment using ``serverless-webpack``
+
+If you are using serverless-webpack, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin to automatically remove ``prisma generate`` and **unnecessary engines** to reduce package capacity.
```
```suggestion
If you are using serverless-webpack, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin to automatically remove `prisma generate` and **unnecessary engines** to reduce package capacity.
```

janpio commented:
Hey @danieluhm2004, interesting plugin! Do I understand correctly that it basically automated (and improves) what we document in https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-aws-lambda#package-pattern-in-serverlessyml?

danieluhm2004 commented:
@janpi, Some are right and some are wrong. Since the package.patterns option is not applied in serverless-webpack, it automatically builds and automatically deletes unnecessary (ex. darwin) binaries when including prisma in serverless-webpack. :)

Due to busy schedules, revisions are being delayed, and we plan to improve the above-mentioned contents a little more within this week.

Thank you.

danieluhm2004 commented:
@janpio , Some are right and some are wrong. Since the package.patterns option is not applied in serverless-webpack, it automatically builds and automatically deletes unnecessary (ex. darwin) binaries when including prisma in serverless-webpack. :)

Due to busy schedules, revisions are being delayed, and we plan to improve the above-mentioned contents a little more within this week.

Thank you!

afcode123 commented:
Hi!
I would be very interested to try this, webpack, lambda and prisma are giving me a headache. When is it available?

janpio commented:
This is just a documentation change @afcode123 - you can just look at the text and do what it says if this works for you.

danieluhm2004 commented:
@janpio Hello :)

I added the document in a little more detail, can you review it again?

janpio commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,55 @@ package:
 
 This ensures the packaged archive is as small as possible.
 
+But, if you are using ``serverless-webpack``, you cannot use that method.
```
```suggestion
But, if you are using `serverless-webpack`, you cannot use that method.
```

ruheni commented:
I have a couple of questions about the guide that came up while following the guide:
- Is there a reason why the workflow defines the db tables using the SQL file, and then run `db pull` to generate the Prisma Schema? I found this a little confusing and we can improve the flow there. I think we can use `prisma migrate dev` here.
- We can also add the copy button for terminal commands and the code blocks for this example.
- Deployment with serverless isn't smooth and experienced some issues deploying the example that I'll look into - pointed out in https://github.com/prisma/prisma-examples/issues/3308.

janpio commented:
Is this in relation to this PR @ruheni? Otherwise optimally open a new issue.

ruheni commented:
Hi @danieluhm2004 

I can't seem to get an example using `serverless-webpack` and `serverless-webpack-prisma` running. I'm running into the following error:

```bash
/bin/sh: prisma: command not found
 
 Error ---------------------------------------------------
 
  Error: Command failed: prisma generate
  /bin/sh: prisma: command not found
  
      at checkExecSyncError (node:child_process:826:11)
      at Object.execSync (node:child_process:900:15)
      at ServerlessWebpackPrisma.onBeforeWebpackPackage (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack-prisma/src/index.js:47:20)
```

Here's the minimal `webpack.config.js` file I'm using:

```js
const path = require("path")
const nodeExternals = require("webpack-node-externals")
const CopyPlugin = require('copy-webpack-plugin')

module.exports = {
  mode: 'development',
  entry: [
    './handlers/users.js',
    './handlers/create-user.js',
    './handlers/seed.js',
    './handlers/status.js',
    './handlers/posts.js',
  ],
  output: {
    // output file for bundled code
    path: path.resolve(__dirname, 'dist'),
    filename: "[name].js"
  },
  // webpack to generate a Node.js bundle
  target: "node",
  externals: [ nodeExternals() ],
  plugins: [
    new CopyPlugin({
      patterns: [
        './prisma/schema.prisma'
      ]
    }),
  ]
}
```

I've also tried it out without `copy-webpack-plugin` but I'm still running into the same error. `prisma generate` isn't being executed, and I'm not sure why. Could you look into this? 

Could you also add a minimal `webpack.config.js` file for the example for folks who would be interested in `serverless-webpack-prisma` and `serverless-webpack`?

danieluhm2004 commented:
> Hi @danieluhm2004
> 
> I can't seem to get an example using `serverless-webpack` and `serverless-webpack-prisma` running. I'm running into the following error:
> 
> ```shell
> /bin/sh: prisma: command not found
>  
>  Error ---------------------------------------------------
>  
>   Error: Command failed: prisma generate
>   /bin/sh: prisma: command not found
>   
>       at checkExecSyncError (node:child_process:826:11)
>       at Object.execSync (node:child_process:900:15)
>       at ServerlessWebpackPrisma.onBeforeWebpackPackage (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack-prisma/src/index.js:47:20)
> ```
> 
> Here's the minimal `webpack.config.js` file I'm using:
> 
> ```js
> const path = require("path")
> const nodeExternals = require("webpack-node-externals")
> const CopyPlugin = require('copy-webpack-plugin')
> 
> module.exports = {
>   mode: 'development',
>   entry: [
>     './handlers/users.js',
>     './handlers/create-user.js',
>     './handlers/seed.js',
>     './handlers/status.js',
>     './handlers/posts.js',
>   ],
>   output: {
>     // output file for bundled code
>     path: path.resolve(__dirname, 'dist'),
>     filename: "[name].js"
>   },
>   // webpack to generate a Node.js bundle
>   target: "node",
>   externals: [ nodeExternals() ],
>   plugins: [
>     new CopyPlugin({
>       patterns: [
>         './prisma/schema.prisma'
>       ]
>     }),
>   ]
> }
> ```
> 
> I'm using Prisma v3.5.0. Node v16.13.0 and npm v8.1.0
> 
> I've also tried it out without `copy-webpack-plugin` but I'm still running into the same error. `prisma generate` isn't being executed, and I'm not sure why. Could you look into this?
> 
> Could you also add a minimal `webpack.config.js` file for the example for folks who would be interested in `serverless-webpack-prisma` and `serverless-webpack`?

Hello.

First of all, thank you so much for your interest in this package.

When I checked, it seems that the error occurred because the executable binary location could not be found because the path environment variable was not specified while prisma was in the dev dependency instead of the global dependency.

I've always used yarn scripts to deploy, so it doesn't seem to have caused the problem by automatically finding the executable binary location.

When executing a command as child_process , it has been modified to automatically find the executable binary location using the npx command, so it would be good to check it again.

thank you.

ruheni commented:
Thank you for the explanation.

>I've always used yarn scripts to deploy

I think it would be reasonable to make it work with npm to reduce the barrier of entry. This could create a barrier of entry for some users who don't have yarn installed.

Could you make the updates:
- Adding a minimal `webpack.config.js` file new users can refer to
- Make it compatible with npm/ npx

danieluhm2004 commented:
> I think it would be reasonable to make it work with npm to reduce the barrier of entry. This could create a barrier of entry for some users who don't have yarn installed.
> 
> Could you make the updates:
> 
> * Adding a minimal `webpack.config.js` file new users can refer to
> * Make it compatible with npm/ npx

Hello.

It has been changed to be supported in all situations with npx.

thank you. :)

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
````diff
@@ -340,6 +340,55 @@ package:
+        - prisma generate
+```
+
+If the setting is complete, check whether the engine is normally deleted and distributed through `serverless` deployment as shown below.
````
```suggestion
Once you've updated your `serverless.yml` file, redeploy your application to check whether the right engine is packaged and distributed by running `serverless deploy` as shown below: 
```

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,55 @@ package:
+
+1. There is no need to separately copy the schema.prisma file via the `webpack plugin`.
+2. There is no need to add `prisma generate` command to webpack options separately.
+3. Except for the Prisma engine used in Lambda, it is automatically deleted. This can save more than 40mb of capacity.
```
Does this mean there is no need for package patterns in `serverless.yml`?

danieluhm2004 replied:
It's a little different.

The file created when serverless-webpack runs the webpack script part is not affected by serverless.patterns.

That is why it is necessary to delete them through this library.

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,55 @@ package:
+
+### Deployment using `serverless-webpack`
+
+If you are using serverless-webpack, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin without any additional configuration.
```
```suggestion
If you are using `serverless-webpack`, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin without any additional configuration.
```

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,55 @@ package:
+If you are using serverless-webpack, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin without any additional configuration.
+
+
+1. There is no need to separately copy the schema.prisma file via the `webpack plugin`.
+2. There is no need to add `prisma generate` command to webpack options separately.
+3. Except for the Prisma engine used in Lambda, it is automatically deleted. This can save more than 40mb of capacity.
```
1. Copies your `schema.prisma` via the Webpack plugin
2. Runs the `prisma generate` command. This means there is no need to add the command to the webpack options separately
3. Packages the correct Prisma engine used in the Lambda. This can save more than 40mb of capacity.

danieluhm2004 replied:
I changed :)

ruheni commented:
Hey @danieluhm2004 👋🏽

>It has been changed to be supported in all situations with npx.

Thanks for the update.  I added a few suggestions on the sections that you updated. Could you add a minimal `webpack.config.js` file for this example? You can place it in the [Expand/ Collapse component](https://www.prisma.io/docs/about/prisma-docs/style-guide/mdx-examples#expandcollapse-section).

Could you also add tests to `serverless-webpack-prisma`? Deployment was successful, but I ran into an Internal Server Error after calling one Lambda function. I think it has something to do with how webpack bundles the handlers.

danieluhm2004 commented:
Hello! @ruheni 

> Thanks for the update. I added a few suggestions on the sections that you updated. Could you add a minimal webpack.config.js file for this example? You can place it in the Expand/ Collapse component.

Since there is no need to separately modify webpack.config.js for this plugin, it is not added separately.

> Could you also add tests to serverless-webpack-prisma? Deployment was successful, but I ran into an Internal Server Error after calling one Lambda function. I think it has something to do with how webpack bundles the handlers.

I thought it was a very important point, so from version 1.0.7, I added a test process using jest when pre-committing.

Please review it again.
thank you.

kihonq commented:
@danieluhm2004 I'm having an error when using the `serverless-webpack-prisma` plugin as shown below:

![image](https://user-images.githubusercontent.com/37780658/147584587-dbfc3cd4-60a9-4b78-ac75-380f70082704.png)


**I have done the following so far**:
1. Remove `package.patterns`

![image](https://user-images.githubusercontent.com/37780658/147584810-5e387705-6ade-4945-a505-51dbac7ff7bc.png)

2. Remove `prisma generate` in `webpack.packageOptions.scripts`

![image](https://user-images.githubusercontent.com/37780658/147584769-397e3063-9ce2-4465-93eb-821792465603.png)
_Note that I use `yarn` for `packager`_

3. Remove usage of `copy-webpack-plugin`

![image](https://user-images.githubusercontent.com/37780658/147584887-819cbc22-2937-4c4e-9649-08c138e98d83.png)


I think it's a good call out from @ruheni to provide us with a minimal example so unrelated causes can be filtered out.
Thanks for the great package !

danieluhm2004 commented:
> @danieluhm2004 I'm having an error when using the `serverless-webpack-prisma` plugin as shown below:
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/37780658/147584587-dbfc3cd4-60a9-4b78-ac75-380f70082704.png)
> 
> 
> 
> 
> 
> **I have done the following so far**:
> 
> 1. Remove `package.patterns`
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/37780658/147584810-5e387705-6ade-4945-a505-51dbac7ff7bc.png)
> 
> 
> 
> 2. Remove `prisma generate` in `webpack.packageOptions.scripts`
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/37780658/147584769-397e3063-9ce2-4465-93eb-821792465603.png)
> 
> _Note that I use `yarn` for `packager`_
> 
> 
> 
> 3. Remove usage of `copy-webpack-plugin`
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/37780658/147584887-819cbc22-2937-4c4e-9649-08c138e98d83.png)
> 
> 
> 
> 
> 
> I think it's a good call out from @ruheni to provide us with a minimal example so unrelated causes can be filtered out.
> 
> Thanks for the great package !
> 
> 

I have found the same issue right now. Clashes with the latest prisma appear to occur. I will correct it as soon as possible.

danieluhm2004 commented:
@kihonq Thank you for your interest in the package.

Just released **1.0.8** and it has been fixed.

Through the recent Prisma update, it has been confirmed that the above error occurs when npx prisma generate is not used in the same location as node_modules. It is predicted that it was probably intended to solve problems arising from other intentions.

However, in the case of serverless-webpack, a virtual project is created and dependencies are installed there. And the package is used by calling prisma cli from node_modules in the project through npx.

So, a crash occurred because the prisma cli in node_modules in the original project was used in the virtual project created by serverless-webpack.
 
If you update, there should be no problems.

thank you.

kihonq commented:
@danieluhm2004 You're right !
Downgraded *prisma* to `3.6.0` fix the issue.
Also I tried your new release `1.0.8` and it works flawlessly with the latest `prisma`.

Thank you for making open source world a better place !

ruheni commented:
> Since there is no need to separately modify webpack.config.js for this plugin, it is not added separately.

The `webpack.config.js` would be helpful for users who are new to `serverless-webpack` and help them set things up without running into any issues.

I was able to deploy the app to AWS, but after trying to call one of the lambda's, I got an HTTP 500 error.

Here's my webpack config:
```js
const path = require("path")
const nodeExternals = require("webpack-node-externals")

module.exports = (async () => {
  return {
    entry: [
      './handlers/users.js',
      './handlers/create-user.js',
      './handlers/seed.js',
      './handlers/status.js',
      './handlers/posts.js',
    ],
    target: 'node',
    externals: [nodeExternals()],
    plugins: [],
    output: {
      libraryTarget: 'commonjs',
      path: path.resolve(__dirname, '.webpack'),
      filename: '[name].js'
    }
  };
})();
```

On running the `sls` locally, I get the following error:

```
webpack compiled with 1 warning in 301 ms
 
 Error ---------------------------------------------------
 
  Error: Cannot find module '/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/.webpack/service/handlers/seed'
  Require stack:
  - /Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/lib/utils.js
  - /Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/lib/validate.js
  - /Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/index.js
  - /Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless/lib/classes/PluginManager.js
  - /Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless/lib/Serverless.js
  - /Users/ruheni/.fnm/node-versions/v16.13.1/installation/lib/node_modules/serverless/lib/Serverless.js
  - /Users/ruheni/.fnm/node-versions/v16.13.1/installation/lib/node_modules/serverless/scripts/serverless.js
  - /Users/ruheni/.fnm/node-versions/v16.13.1/installation/lib/node_modules/serverless/bin/serverless.js
      at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
      at Function.resolve (node:internal/modules/cjs/helpers:108:19)
      at searchAndProcessCache (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/lib/utils.js:34:25)
      at Object.purgeCache (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/lib/utils.js:21:10)
      at ServerlessWebpack.prepareLocalInvoke (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/serverless-webpack/lib/prepareLocalInvoke.js:33:18)
      at ServerlessWebpack.tryCatcher (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/promise.js:547:31)
      at Promise._settlePromise (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/promise.js:604:18)
      at Promise._settlePromise0 (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/promise.js:649:10)
      at Promise._settlePromises (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/promise.js:729:18)
      at _drainQueueStep (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/async.js:93:12)
      at _drainQueue (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/async.js:86:9)
      at Async._drainQueues (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/async.js:102:5)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/ruheni/Documents/repos/work/prisma/temp/prisma-aws-lambda/node_modules/bluebird/js/release/async.js:15:14)
      at processImmediate (node:internal/timers:464:21)
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              16.13.1
     Framework Version:         2.69.1 (local)
     Plugin Version:            5.5.1
     SDK Version:               4.3.0
     Components Version:        3.18.1
```
I have a feeling it's something to do with my webpack configuration. I'm not familiar with webpack.

I'm running the latest versions of Prisma, serverless-webpack-prisma, and serverless-webpack.

I think once this is resolved, we can merge the PR.

danieluhm2004 commented:
@ruheni Hello.

We have added a setting method that can be set quickly for users who do not have webpack installed.

Thank you for checking it out.

ruheni commented:
Hey @danieluhm2004

Thanks for adding the starter webpack configuration files. 

I'm still experiencing some issues invoking the functions. I keep running into an Internal server error after deploy. 

I ran `sls logs -f seed -t` to watch the logs when I invoked my lambda and here's the error that I got:

```bash
2022-01-14 12:39:35.089 (+01:00)        undefined       ERROR   Uncaught Exception      {"errorType":"Runtime.ImportModuleError","errorMessage":"Error: Cannot find module '@prisma/client'\nRequire stack:\n- /var/task/handlers/seed.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js","stack":["Runtime.ImportModuleError: Error: Cannot find module '@prisma/client'","Require stack:","- /var/task/handlers/seed.js","- /var/runtime/UserFunction.js","- /var/runtime/index.js","    at _loadUserApp (/var/runtime/UserFunction.js:100:13)","    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:999:30)","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)","    at Module.load (internal/modules/cjs/loader.js:863:32)","    at Function.Module._load (internal/modules/cjs/loader.js:708:14)","    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)","    at internal/main/run_main_module.js:17:47"]}
START RequestId: 81f5bb31-5bf1-4bb7-9b0d-a5309a076f3b Version: $LATEST
2022-01-14 12:39:35.320 (+01:00)        undefined       ERROR   Uncaught Exception      {"errorType":"Runtime.ImportModuleError","errorMessage":"Error: Cannot find module '@prisma/client'\nRequire stack:\n- /var/task/handlers/seed.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js","stack":["Runtime.ImportModuleError: Error: Cannot find module '@prisma/client'","Require stack:","- /var/task/handlers/seed.js","- /var/runtime/UserFunction.js","- /var/runtime/index.js","    at _loadUserApp (/var/runtime/UserFunction.js:100:13)","    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:999:30)","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)","    at Module.load (internal/modules/cjs/loader.js:863:32)","    at Function.Module._load (internal/modules/cjs/loader.js:708:14)","    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)","    at internal/main/run_main_module.js:17:47"]}
```

The deployment logs indicate that the Prisma Client was generated on running `sls deploy`:
```
Serverless: **Install prisma devDependencies for generate**
Serverless: **Copy prisma schema for service**...
Serverless: **Generate prisma client for service**...
Serverless: Remove unused prisma engine:
Serverless: - node_modules/.prisma/client/libquery_engine-darwin.dylib.node
Serverless: Remove prisma devDependencies
Serverless: Copying existing artifacts...
Serverless: Packaging service...
Serverless: Ensuring that deployment bucket exists
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service prisma-aws-lambda-example.zip file to S3 (15.88 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
```

This is my plugin setup in `serverless.yml`:
```
plugins:
  - serverless-dotenv-plugin
  - serverless-webpack
  - serverless-webpack-prisma
  ```

I'm running the latest versions of all the packages used in the guide and your library. Could you double-check from your end whether you're experiencing this error too and that Prisma Client is generated?

I used the `aws-lambda` prisma-example and made the necessary modifications according to the guide to deploy the application.

janpio reviewed (changes requested):
Blocking from accidental merge: See ongoing discussion with @ruheni

danieluhm2004 commented:
> Hey @danieluhm2004
> 
> Thanks for adding the starter webpack configuration files.
> 
> I'm still experiencing some issues invoking the functions. I keep running into an Internal server error after deploy.
> 
> I ran `sls logs -f seed -t` to watch the logs when I invoked my lambda and here's the error that I got:

Hello @ruheni!

First of all, thank you for taking the time to do your best for this pull request, although it is lacking a lot.

I tried again according to the documentation I updated, but the issue you mentioned does not occur, so it is not easy to determine the cause.

Could you please send me the project you are testing? You can email it or post it on github and give it a link.

If you send it to me, I think I can check and correct in more detail which parts of the document have led to misleading users.

Thank you so much always!!

ruheni commented:
Hey @danieluhm2004 

Here's the [repository](https://github.com/ruheni/repro-serverless-prisma-aws-lambda) you can use to reproduce the error I keep running into.

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
````diff
@@ -340,6 +340,115 @@ package:
+In the case of TypeScript, we also insert the following configuration inside `exports`:
+
+```javascript
+...skip
+module.exports = {
+  ...skip
+  module: {
+    rules: [
+      {
+        test: /\.tsx?$/,
+        loader: 'ts-loader',
+        exclude: /node_modules/,
+      },
+    ],
+  }
+  ...skip
````
```suggestion
// ...imports
module.exports = {
  // ...initial config
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        loader: 'ts-loader',
        exclude: /node_modules/,
      },
    ],
  }
  // ... additional config
```

danieluhm2004 commented:
Hello @ruheni !

Thank you so much for sending us the example project that doesn't work. I checked what parts were lacking and made additional corrections to the document.

1. I have identified an issue that requires additional setting of `optimization: { concatenateModules: false }` in webpack.config.js.

2. Confirmed that the value of `custom.webpack.includeModules` in serverless.yml must be true.

3. Edited to mention once again that package.patterns needs to be removed.

4. Added that serverless is required for devDependency.

It was a good opportunity to check common cases in detail. In addition, I have put the modifications in the repro project that you sent me as a pull request, so it would be good to check it.

Please check again.
thank you!

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,132 @@ package:
+
+<details><summary>If you are not using webpack yet</summary>
+
+Install packages to use `webpack`.
```
```suggestion
Install dependencies.
```

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
````diff
@@ -340,6 +340,132 @@ package:
+};
+```
+
+Various settings can be checked in [Serverless Webpack](https://www.serverless.com/plugins/serverless-webpack).
````
```suggestion
Refer to the [Serverless Webpack documentation](https://www.serverless.com/plugins/serverless-webpack) for additional configuration.
```

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
````diff
@@ -340,6 +340,132 @@ package:
+
+`serverless-webpack-prisma` automatically runs `prisma generate` so you don't need to use the webpack script option separately as shown below.
+
+```yaml file=serverless.yaml
````
```suggestion
```yaml file=serverless.yml
```
Just a tiny nit pick – let's stick to `serverless.yml` to maintain consistency in the file naming :)

ruheni commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
````diff
@@ -340,6 +340,132 @@ package:
+npm install -D serverless-webpack-prisma serverless
+```
+
+Add the corresponding plugin under the `serverless-webpack` plugin as shown below.
````
```suggestion
Update your `serverless.yml` by adding the `serverless-webpack` to the plugins list. 
```

nikolasburk commented on content/300-guides/200-deployment/110-deployment-guides/400-deploying-to-aws-lambda.mdx:
```diff
@@ -340,6 +340,132 @@ package:
+
+If you are using `serverless-webpack`, you can use the [serverless-webpack-prisma](https://github.com/danieluhm2004/serverless-webpack-prisma) plugin without any additional configuration. `serverless-webpack-prisma` does the following:
+
+1. Copies your schema.prisma via the Webpack plugin
```
```suggestion
1. Copies your `schema.prisma` via the Webpack plugin
```

danieluhm2004 commented:
Sounds great.
I applied all of the suggestions.

Thank you.

danieluhm2004 commented:
Hello?

Are you all done? When are they merged?

And can I use the all-contributors command to add to all-contributors?

thank you.

tanberry commented:
Hullo and thank you @danieluhm2004 for your contribution! Apologies that it is just now being merged. Thanks again for your help.
